Cpp/odin/facebook.cpp

```diff
@@ -15,67 +15,125 @@
 
 #include <fost/insert>
 
-
-f5::u8string odin::facebook::get_app_token(
-        const f5::u8string &app_id, const f5::u8string &app_secret) {
-    fostlib::url base_url(odin::c_facebook_endpoint.value());
-    fostlib::url::filepath_string api{"/oauth/access_token"};
-    fostlib::url::query_string qs{};
-    fostlib::url fb_url(base_url, api);
-    qs.append("client_id", app_id);
-    qs.append("client_secret", app_secret);
-    qs.append("grant_type", "client_credentials");
-    fb_url.query(qs);
-    fostlib::http::user_agent ua(fb_url);
-    auto response = ua.get(fb_url);
-    auto response_data = fostlib::coerce<fostlib::string>(
-            fostlib::coerce<fostlib::utf8_string>(response->body()->data()));
-    fostlib::json body = fostlib::json::parse(response_data);
-    return fostlib::coerce<f5::u8string>(body["access_token"]);
-}
+namespace {
 
 
-bool odin::facebook::is_user_authenticated(
-        f5::u8view app_token, f5::u8view user_token) {
-    fostlib::url base_url(odin::c_facebook_endpoint.value());
-    fostlib::url::filepath_string api{"/debug_token"};
-    fostlib::url fb_url(base_url, api);
-    fostlib::url::query_string qs{};
-    qs.append("input_token", user_token);
-    qs.append("access_token", app_token);
-    fb_url.query(qs);
-    fostlib::http::user_agent ua(fb_url);
-    auto response = ua.get(fb_url);
-    auto response_data = fostlib::coerce<fostlib::string>(
-            fostlib::coerce<fostlib::utf8_string>(response->body()->data()));
-    fostlib::json body = fostlib::json::parse(response_data);
-    return fostlib::coerce<bool>(body["data"]["is_valid"]);
+    std::unique_ptr<fostlib::http::user_agent::response> get_or_mock(
+            fostlib::http::user_agent &ua,
+            fostlib::url url,
+            fostlib::json config = {}) {
```

Comment:
If you want to set a default I think you should do it in the header, not here.

Reply:
I defined this function in an anonymous namespace, this also doesn't include in the header as well.

Reply:
Oh, OK. That's fine then. It does look odd though. When I first read it I thought "empty object", but of course it's the default constructor which is `null`. `fostlib::null` might be a clearer value to use